File: LinkedList/460. LFU Cache/Solution.py
class Node:
    def __init__(self, val = 0, prev = None, next = None):
        self.val = val
        self.prev = prev
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = self.tail = None
        self.nodeMap = {}

    def length(self) -> int:
        return len(self.nodeMap)
    
    def pushLeft(self, val: int) -> None:
        if val in self.nodeMap:
            return None
        
        new_node = Node(val)
        self.nodeMap[val] = new_node
        
        if self.head == None:
            self.head = self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = self.head.prev
    
    def pushRight(self, val: int) -> None:
        if val in self.nodeMap:
            return None

        new_node = Node(val)
        self.nodeMap[val] = new_node
        
        if self.tail == None:
            self.tail = self.head = new_node
        else:
            self.tail.next = new_node
            new_node.prev = self.tail
            self.tail = self.tail.next
    
    def popleft(self) -> int:
        val = self.pop(self.head.val)
        return val

    def popRight(self) -> int:
        val = self.pop(self.tail.val)
        return val
    
    def pop(self, val) -> int:
        if self.length() == 0:
            return -1

        curr = self.nodeMap.pop(val)
        if self.head == self.tail == curr:
            self.head = self.tail = None
        elif self.head == curr:
            self.head =  self.head.next
            self.head.prev = None
        elif self.tail == curr:
            self.tail = self.tail.prev
            self.tail.next = None
        else:
            curr.next.prev = curr.prev
            curr.prev.next = curr.next
        return curr.val

class LFUCache:
    # N : total no of nodes
    # Time Complexity: O(1)
    # Space Complexity: O(N)
    def __init__(self, capacity: int):
        self.MAX_LIMIT = capacity
        self.lfu = 0
        self.keyMap = {}
        self.countMap = {}

    def updateCounter(self, key: int) -> None:
        curr_count = self.keyMap[key][1]
        self.countMap[curr_count].pop(key)
        if (curr_count + 1) not in self.countMap:
                self.countMap[curr_count + 1] = LinkedList()
        self.countMap[curr_count + 1].pushLeft(key)
        self.keyMap[key][1] += 1
        
        if self.countMap[curr_count].length() == 0 and self.lfu == curr_count:
            self.lfu += 1


    def get(self, key: int) -> int:
        if key not in self.keyMap:
            return -1
        self.updateCounter(key)
        return self.keyMap[key][0]

    def put(self, key: int, value: int) -> None:
        if key in self.keyMap:
            self.keyMap[key][0] = value
            self.updateCounter(key)
        else:
            if len(self.keyMap) == self.MAX_LIMIT:
                self.keyMap.pop(self.countMap[self.lfu].popRight())
            self.keyMap[key] = [value, 1]
            if 1 not in self.countMap:
                self.countMap[1] = LinkedList()
            self.countMap[1].pushLeft(key)
            self.lfu = 1

File: LinkedList/460. LFU Cache/Solution.java
import java.util.*;

class Node {
    public int val;
    public Node prev;
    public Node next;

    public Node(int val) {
        this.val = val;
        this.prev = this.next = null;
    }
}

class LinkedList1 {
    private Node head;
    private Node tail;
    private Map<Integer, Node> nodeMap;

    public LinkedList1() {
        this.nodeMap = new HashMap<>();
        this.head = this.tail = null;
    }

    public int length() {
        return this.nodeMap.size();
    }
    
    public void pushLeft(int val) {
        if (nodeMap.containsKey(val)) return;
        Node new_node = new Node(val);
        nodeMap.put(val, new_node);

        if (head == null) head = tail= new_node;
        else {
            head.prev = new_node;
            new_node.next = head;
            head = head.prev;
        }
    }

    public void pushRight(int val) {
        if (nodeMap.containsKey(val)) return;
        Node new_node = new Node(val);
        nodeMap.put(val, new_node);

        if (tail == null) head = tail = new_node;
        else {
            tail.next = new_node;
            new_node.prev = tail;
            tail = tail.next;
        }
    }

    public int pop(int val) {
        if (length() == 0) return -1;

        Node curr = nodeMap.remove(val);
        if (head == tail & tail == curr) {
            head = tail = null;
        } else if (head == curr) {
            head = head.next;
            head.prev = null;
        } else if (tail == curr) {
            tail = tail.prev;
            tail.next = null;
        } else {
            curr.next.prev = curr.prev;
            curr.prev.next = curr.next;
        }
        return curr.val;
    }

    public int popLeft() {
        return pop(head.val);
    }

    public int popRight() {
        return pop(tail.val);
    }
    
}

class LFUCache {
    // N : total no of nodes
    // Time Complexity: O(1)
    // Space Complexity: O(N)
    private static int MAX_LIMIT;
    private Map<Integer, Integer[]> keyMap;
    private Map<Integer, LinkedList1> countMap;
    private int lfu;
    public LFUCache(int capacity) {
        MAX_LIMIT = capacity;
        keyMap = new HashMap<>();
        countMap = new HashMap<>();
        lfu = 0;
    }

    private void updateCounter(int key) {
        int count = keyMap.get(key)[1];
        keyMap.put(key, new Integer[] {keyMap.get(key)[0], count + 1});

        countMap.get(count).pop(key);
        if (!countMap.containsKey(count + 1)) countMap.put(count + 1, new LinkedList1());
        countMap.get(count + 1).pushLeft(key);
        if (lfu == count && countMap.get(count).length() == 0) lfu++;
    }
    
    public int get(int key) {
        if (keyMap.size() == 0 || !keyMap.containsKey(key)) return -1;
        updateCounter(key);
        return keyMap.get(key)[0];
    }
    
    public void put(int key, int value) {
        if (keyMap.containsKey(key)) {
            keyMap.get(key)[0] = value;
            updateCounter(key);
        } else {
            if (keyMap.size() == MAX_LIMIT) {
                keyMap.remove(countMap.get(lfu).popRight());
            }
            keyMap.put(key, new Integer[]{value, 1});
            if (!countMap.containsKey(1)) countMap.put(1, new LinkedList1());
            countMap.get(1).pushLeft(key);
            lfu = 1;
        }
    }
}